### ISSUE TITLE
Transformation of weather maps to Mercator projection

### ISSUE THREAD

**Issue by mikebravoyes** [osmdroid/osmdroid#1573]
## Issue Type

[x] Question

## Description and/or steps/code to reproduce the problem

I have written an app with a map component and now I want to overlay some weather images, e.g. from [sat24](https://en.sat24.com). The basic osmdroid map is available in Mercator projection; unfortunately the overlaid weather maps are not.

For the combination of both components, I use the class GroundOverlay, which allows defining the four corner points in lat/lon, even for irregular quadrangles. For some of the satellite images (for example [Australia](https://en.sat24.com/image?type=visual&region=au)), that fits quite well. 

![Screenshot_australia](https://user-images.githubusercontent.com/6436590/80690838-b05d7680-8acf-11ea-9747-124903c5fa35.jpg)

But some of the images (e.g. the region [Japan+Korea](https://en.sat24.com/image?type=visual&region=jp)) are extremely distorted, so that a linear transformation has no chance for a reasonable match.

![Screenshot_japan+korea](https://user-images.githubusercontent.com/6436590/80690862-b7848480-8acf-11ea-9c68-db6dfba8df9a.jpg)

Are there any classes available, which manage a transformation from a satellite image with an unknown projection to a Mercator map? I think about something like a morphing algorithm with a lot of (much more than 4) control points.

### Version of osmdroid the issue relates to:

6.1.6

**Comment by monsieurtanuki** [comment]
I would find inspiration in [`GroundOverlay4`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/GroundOverlay4.java) and build the corresponding `Matrix`'s.

**Comment by mikebravoyes** [comment]
@monsieurtanuki 
Hmm, the docu says: GroundOverlay4 is deprecated, use GroundOverlay instead. (That's what i do already.) Both classes consider just the 4 corner points of a quadrangle, not more.

**Comment by monsieurtanuki** [comment]
Let me rephrase it: I would find inspiration in [`GroundOverlay`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/GroundOverlay.java) and build the corresponding `Matrix`'s.
If I remember well the idea is to have an input matrix and an output matrix. Currently we use 4 points.
Don't be shy: create a new class but with a much bigger matrix. Wouldn't that cover your needs?

**Comment by mikebravoyes** [comment]
@monsieurtanuki 
Yes, probably. But the devil is in the details. It's probably a high sophisticated algorithm required for that. My hope was, that somebody already had done it, so I have not to reinvent the wheel.

**Comment by mikebravoyes** [comment]
@monsieurtanuki 
I had a deeper look into GroundOverlay and found that the real transformation is invoked by 
`Matrix.setPolyToPoly(mMatrixSrc, 0, mMatrixDst, 0, 4);`
(Please note, that the number of matrix points is set to constant 4, which is checked during the further processing. Numbers > 4 lead to an IllegalArgumentException.)

And in the class Matrix (package android.graphics) we end up with 
```
@FastNative
private static native boolean nSetPolyToPoly(long nObject,
    float[] src, int srcIndex, float[] dst, int dstIndex, int pointCount);
```
Which means, the real transformation is done in native code for performance reasons. Thus, even if I would invent a high sophisticated algorithm for more than 4 points in Java, it would probably be too slow for a reasonable usage. What a pity!

**Comment by monsieurtanuki** [comment]
Oh, I assumed you could put more than 4 points in `Matrix.setPolyToPoly`. Fair enough.

What's wrong exactly with the Japan map? You say that it's very distorted but I still see a clean quadrilateral: a 4-point Matrix conversion should be enough, shouldn't it? Is it too big to fit in a "standard" tile?

**Comment by mikebravoyes** [comment]
@monsieurtanuki 
No, it's not the size. Look at the yellow lat/lon grid in the Japan map. The latitudes are curved and the longitudes are converging. In the osmdroid background map the lat and lon lines are perpendicular to each other - as it should be in a Mercator projection.

The two projections are simply not compatible. The Japan image can't be transformed to a perfect match by a simple 4-point matrix conversion.

My idea was, to achieve a half-way usable match with something like a morphing algorithm. Obviously, it's not as easy as I thought.

**Comment by spyhunter99** [comment]
May want to look into gdal. I'm pretty sure it can be compiled for
android/arm and i can do reprojections based on set of points and their
corresponding lat/lons

**Comment by monsieurtanuki** [comment]
@mikebravoyes Now I understand - actually it didn't look that bad. But you're right, the yellow lines don't look good.
I would split the input file into smaller chunks that can be converted by 4-point matrices without too much damage.

**Comment by mikebravoyes** [comment]
@spyhunter99 
I just read about GDAL. Seems to be a very mighty tool.
